Writable JSON API changes

kasiabulat opened the pull request:
I added majority of API and implementation changes and missing functionalities, including:
* Changing null `JsonNode` into `JsonNull`
* Moving implicit casts operators to `JsonNode`
* Rewriting `JsonObject` to use a combination of dictionary and linked list of properties
* Adding my own version of `Parse`
* Moving handling duplicates from `JsonObject` to `Parse`
* Rewriting `DeepCopy` not to use recursion
* Adding `WriteTo(Utf8JsonWriter)`  
* Adding `ToJsonString` using `WriteTo(Utf8JsonWriter)`  

I am still missing adding `StringComparison` to methods looking up properties. I didn't change the naming (to include "tree" prefix) for now. And I need to add much more tests, especially testing new features.

addresses: #39922
cc: @bartonjs @ahsonkhan @ericstj @joperezr @terrajobst @JamesNK @stephentoub

kasiabulat committed:
minor changes

kasiabulat committed:
implicit operators moved to JsonNode

kasiabulat committed:
JsonNull introduced

kasiabulat committed:
work on rewriting JsonObject to use JsonObjectProperty

kasiabulat committed:
minor fixes

kasiabulat committed:
JsonElement.ObjectEnumerator changed to use _current instead of _enumerator

kasiabulat committed:
enumerator fixes

kasiabulat committed:
iterative DeepCopy implemented

kasiabulat committed:
minor fixes, more complex deep copy test added

kasiabulat committed:
code reorganized

kasiabulat committed:
work on parse

kasiabulat committed:
parse implemented

kasiabulat committed:
parse fixes

kasiabulat committed:
JsonReaderOptions added to JsonNode.Parse

kasiabulat committed:
WriteTo and ToJsonString added

kasiabulat committed:
ToJsonString fixed

kasiabulat committed:
work on fixing WriteTo to handle duplicate properties in different objects

kasiabulat committed:
WriteTo duplicate properties in different objects fixed

kasiabulat committed:
WriteTo fixes

kasiabulat committed:
code refactored

kasiabulat committed:
specification adjusted

kasiabulat committed:
csproj fixes

kasiabulat commented:
I am not sure if naming and organization of files is OK. I wanted to join all of the methods performing JSON element/node/string traversal into `Json.Node` parsing, because they have similar logic. Probably I could try to unify more the way they handle end of object/array and how they retrieve additional information from stack.

kasiabulat commented:
Of course I hoped everything will work on netfx, I will fix failure when applying review feedback.

kasiabulat committed:
work on string comparison

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonObject.cs:
```diff
@@ -72,7 +70,7 @@ public JsonObject(DuplicatePropertyNameHandling duplicatePropertyNameHandling =
         /// <param name="jsonProperty">The property to add.</param>
         /// <exception cref="ArgumentException">
-        ///   Property name to set already exists if handling duplicates is set to <see cref="DuplicatePropertyNameHandling.Error"/>.
+        ///   Property name to set already exists.
```
```suggestion
        ///   Property name to add already exists.
```

(I only noticed this because you fixed the next one, but there are others that got missed).

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonObject.cs:
```diff
@@ -378,8 +164,9 @@ public void Add(string propertyName, IEnumerable<JsonNode> propertyValues)
             var jsonArray = new JsonArray();
```
I guess the purpose of this method is so you can do something like `obj.Add("numbers", arr.OfType<JsonNumber>())`?

Is that common enough, useful enough, and discoverable enough, that we want the method here instead of just getting the user to wrap their enumerable in `new JsonArray(...)`?

kasiabulat replied:
Yes, it's just for convenience. I can remove it for now.

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonObject.cs:
```diff
@@ -378,8 +164,9 @@ public void Add(string propertyName, IEnumerable<JsonNode> propertyValues)
             {
-                jsonArray.Add(value);
+                jsonArray.Add(value ?? new JsonNull());
             }
```
Isn't there a JsonArray ctor that takes an enumerable of nodes? Why doesn't this method just call that instead of repeating the work?

kasiabulat replied:
Right. I will remove it anyway.

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonObject.cs:
```diff
@@ -394,7 +181,46 @@ public void Add(string propertyName, IEnumerable<JsonNode> propertyValues)
+                throw new ArgumentNullException(nameof(propertyName));
+            }
+
+            if (!_dictionary.ContainsKey(propertyName))
```
This method is currently "tripple-dipping" the dictionary.  ContainsKey, get_Item, and Remove.

They can all be combined into

```C#
if (!_dictionary.Remove(propertyName, out JsonObjectProperty value))
```

If it returns false, there was no match.  If it returns true, there's now no match and you have the existing value to manipulate the linked list.

kasiabulat replied:
Thanks, I wasn't aware of this method! And I am rewriting it anyway a little bit now, because of adding `StringComparison`.

kasiabulat replied:
So this `Remove` does not work for netfx :(

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonObject.cs:
```diff
@@ -534,14 +369,14 @@ public bool TryGetJsonArrayPropertyValue(string propertyName, out JsonArray json
         ///  A collection containing the property values of JSON object.
         /// </summary>
-        public ICollection<JsonNode> PropertyValues => _dictionary.Values;
+        public ICollection<JsonNode> PropertyValues => _dictionary.Values.Select(jsonObjectProperty => jsonObjectProperty._value).ToList();
```
Oh, I missed this earlier.  We usually don't let properties return interfaced collections, unless they got their collection as an interfaced ctor parameter (or are settable).

This should possibly say List; but it's expensive, so it shouldn't be a property.  (Dictionary got this wrong).

kasiabulat replied:
Ok, I see, I'll change it to method.

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonObject.cs:
```diff
@@ -550,7 +385,7 @@ public bool TryGetJsonArrayPropertyValue(string propertyName, out JsonArray json
         /// <returns>An enumerator structure for the JSON object.</returns>
         /// <exception cref="ArgumentException">
-        ///   Property name to set already exists if handling duplicates is set to <see cref="DuplicatePropertyNameHandling.Error"/>.
+        ///   Property name to set already exists.
```
Why does GetEnumerator set properties? :)

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonObjectEnumerator.cs:
```diff
@@ -8,38 +8,63 @@ namespace System.Text.Json
         ///   Gets the property in the JSON object at the current position of the enumerator.
         /// </summary>
-        public KeyValuePair<string, JsonNode> Current => _enumerator.Current;
+        public KeyValuePair<string, JsonNode> Current => new KeyValuePair<string, JsonNode>(_current._name, _current._value);
```
should this `return default` when _current is null? (before first MoveNext, after last one?)

kasiabulat replied:
Ok, I will fix it.

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonObjectProperty.cs:
```diff
@@ -0,0 +1,22 @@
+        internal string _name { get; set; }
+        internal JsonNode _value { get; set; }
+        internal JsonObjectProperty _prev { get; set; }
+        internal JsonObjectProperty _next { get; set; }
```
These are named like fields, but are properties.  PascalCase all of them, please.

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonObjectProperty.cs:
```diff
@@ -0,0 +1,22 @@
+{
+    internal class JsonObjectProperty
+    {
+        internal string _name { get; set; }
```
I assume Name is intended to be immutable for a given property-value, so it should probably be 

```suggestion
        internal string Name { get; }
```

JamesNK replied:
readonly?

bartonjs commented on src/System.Text.Json/tests/JsonNullTests.cs:
```diff
@@ -0,0 +1,30 @@
+            Assert.IsType<JsonNull>(jsonObject["null3"]);
+            Assert.IsType<JsonNull>(jsonObject["null4"]);
+            Assert.IsType<JsonNull>(jsonObject["null5"]);
+        }
```
Presumably we can also test that the ValueKind on it is Null...  and the ToString...

bartonjs replied:
```C#
JsonNull clone = jsonNull.Clone();
Assert.IsNotSame(clone, jsonNull);
```

Or IsSame, whichever way we want to go.

kasiabulat replied:
Ok, sure, I am working on adding more tests.

JamesNK commented on src/System.Text.Json/tests/JsonObjectTests.cs:
```diff
@@ -310,7 +291,7 @@ public static void TestAddingJsonArrayFromJsonNumberArray()
         {
             var preferences = new JsonObject()
```
Add a test for creating JsonObject using this initializer syntax:

```csharp
var o = new JsonObject
{
    ["hello"] = "world",
    ["boolean"] = true
};
```

JamesNK commented on src/System.Text.Json/src/System/Text/Json/Node/JsonObject.cs:
```diff
@@ -13,39 +14,25 @@ namespace System.Text.Json
-
-            _dictionary = new Dictionary<string, JsonNode>();
-            _duplicatePropertyNameHandling = duplicatePropertyNameHandling;
+            _dictionary = new Dictionary<string, JsonObjectProperty>();
```
Should StringComparison.Ordinal be passed to the dictionary? I think it is the default but being explicit can't hurt

kasiabulat replied:
Ok, I will check it. I am now working on adding `StringComparison`.

kasiabulat replied:
The default comparer for dictionary keys is the equals method for key - string, which performs an ordinal (case-sensitive and culture-insensitive) comparison, so this should be what we want.

JamesNK commented on src/System.Text.Json/ref/System.Text.Json.cs:
```diff
@@ -270,9 +235,43 @@ public abstract partial class JsonNode
         public static System.Text.Json.JsonNode DeepCopy(System.Text.Json.JsonElement jsonElement) { throw null; }
         public static System.Text.Json.JsonNode GetNode(System.Text.Json.JsonElement jsonElement) { throw null; }
-        public static System.Text.Json.JsonNode Parse(string json) { throw null; }
+        public static implicit operator System.Text.Json.JsonNode (bool value) { throw null; }
```
What about nullable types?

```csharp
bool? nullableBool = true;
JsonToken token = nullableBool;
```

Should have implicit operator for all structs:
```csharp
public static implicit operator System.Text.Json.JsonNode (bool? value) { throw null; }
```

ahsonkhan replied:
Let's consider that separately (worth noting in the doc).

JamesNK commented:
Any reason `#nullable enable` isn't set for this? IMO it can help identify issues with an API. For example, using null instead of `JsonNull` would have identified that every `JsonToken` returned or given to the API could be nullable.

kasiabulat committed:
work on string comparison

bartonjs commented:
> Any reason #nullable enable isn't set for this? 

We're starting at the bottom, and working our way up systematically.  (Libraries are being done all at once, the one sub-feature that had nullability already done had it removed recently, because it threw off the progress tracking).

kasiabulat commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNode.Parsing.cs:
```diff
@@ -0,0 +1,323 @@
+        /// </summary>
+        /// <param name="jsonElement"><see cref="JsonElement"/> to copy.</param>
+        /// <returns><see cref="JsonNode"/>  representing <paramref name="jsonElement"/>.</returns>
+        public static JsonNode DeepCopy(JsonElement jsonElement)
```
I'll change it to "Traversal".

kasiabulat commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNode.cs:
```diff
@@ -57,67 +55,145 @@ public static bool TryGetNode(JsonElement jsonElement, out JsonNode jsonNode)
-        /// <returns><see cref="JsonNode"/> which is a clone of this instance.</returns>
-        public abstract JsonNode Clone();
+        /// <param name="value">The value to convert.</param>
+        public static implicit operator JsonNode(ReadOnlySpan<char> value) => new JsonString(value);
```
Ok.

kasiabulat commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNode.cs:
```diff
@@ -57,67 +55,145 @@ public static bool TryGetNode(JsonElement jsonElement, out JsonNode jsonNode)
+        /// </summary>
+        /// <param name="value">The value to convert.</param>
+        /// <exception cref="ArgumentException">
+        ///   Provided value is not in a legal JSON number format.
```
Right, now it won't throw, I'll remove it.

kasiabulat committed:
more string comparison tests added

kasiabulat committed:
review comments included

kasiabulat committed:
netfx fixes

kasiabulat committed:
work on tests improvement

joperezr commented on src/System.Text.Json/docs/writable_json_dom_spec.md:
```diff
@@ -93,7 +94,7 @@ var preferences = new JsonObject()
     { "colours", new JsonArray { "red", "green", "purple" } },
     { "numbers", new JsonArray { 4, 123, 88 } },
-    { "prime numbers", new JsonNumber[] { 19, 37 } },
+    { "prime numbers", new JsonNode[] { 19, 37 } },
```
Will the other way still work though? we probably want to support both JsonNode[] and JsonNumber[]

kasiabulat replied:
Actually, @bartonjs sugessted removing IEnumerable ctor on JsonObject at all: https://github.com/dotnet/corefx/pull/40673#discussion_r319272714

joperezr commented on src/System.Text.Json/ref/System.Text.Json.cs:
```diff
@@ -108,7 +74,6 @@ public sealed partial class JsonBoolean : System.Text.Json.JsonNode, System.IEqu
         public bool Equals(System.Text.Json.JsonBoolean other) { throw null; }
         public override int GetHashCode() { throw null; }
         public static bool operator ==(System.Text.Json.JsonBoolean left, System.Text.Json.JsonBoolean right) { throw null; }
-        public static implicit operator System.Text.Json.JsonBoolean (bool value) { throw null; }
```
Is there a downside of keeping this as well? It will be a little weird to have:
```c#
JsonNode aBoolean = false; // Works just fine, and aBoolean is a JsonBoolean
JsonBoolean anotherBoolean = true; // Fails because we don't have implicit converters.
```

kasiabulat replied:
Then probably I would need to add implicit conversions for JsonNumber as well and there is this problematic issue of throwing for values like float.PositiveInfinity (currently I am creating JsonString in such case for JsonNode implicit cast operator).

ahsonkhan replied:
We should document that reasoning for why we don't have the implicit casts on the concrete types in the design doc. Something worth re-visiting in the future.

kasiabulat committed:
more tests added, minor fix

kasiabulat committed:
specification fix

kasiabulat committed:
naming fix

kasiabulat commented:
I addressed review comments, added more tests and StringComparison overloads.

kasiabulat committed:
csproj fix

kasiabulat committed:
work on using JsosReadonlyDomWriteTests

kasiabulat committed:
Merge remote-tracking branch 'upstream/master' into kasiabulat/writable-json-api-changes

kasiabulat committed:
de-scape fix

kasiabulat committed:
work on improving code coverage

kasiabulat commented:
Can I merge this?

kasiabulat committed:
JsonArray(Enumerator) and JsonObject(Enumerator) 100% coverage achieved

kasiabulat committed:
JsonNull and JsonString 100% coverage achieved

kasiabulat committed:
work on improving code coverage

kasiabulat committed:
JsonNode all reachable lines covered

kasiabulat committed:
work on improving JsonElement coverage

kasiabulat committed:
Merge with master

kasiabulat commented:
In the meantime I added more tests to improve code coverage.

JamesNK commented on src/System.Text.Json/src/System/Text/Json/Node/JsonObject.cs:
```diff
@@ -427,20 +302,76 @@ public JsonNode GetPropertyValue(string propertyName)
-        ///   Null <paramref name="jsonNode"/> doesn't mean the property value was "null" unless <see langword="true"/> is returned.
         /// </remarks>
-        public bool TryGetPropertyValue(string propertyName, out JsonNode jsonNode) => _dictionary.TryGetValue(propertyName, out jsonNode);
+        public bool TryGetPropertyValue(string propertyName, out JsonNode jsonNode, StringComparison stringComparison)
```
out parameter should be last

kasiabulat replied:
Sure, thanks, I will change it.

JamesNK commented on src/System.Text.Json/src/System/Text/Json/Node/JsonObject.cs:
```diff
@@ -526,41 +523,54 @@ public bool TryGetJsonArrayPropertyValue(string propertyName, out JsonArray json
         ///   A collection containing the property names of JSON object.
         /// </summary>
-        public ICollection<string> PropertyNames => _dictionary.Keys;
+        public ICollection<string> GetPropertyNames() => _dictionary.Keys;
```
Hmmm, I wonder if `IEnumerable<string>` or `IReadOnlyCollection<string>` would be better here. We don't want people to think they can modify the `JsonObject` via this collection.

kasiabulat replied:
IDictionary properties `Keys` and `Values` have this signature, but as we are not implementing the interface and decided to make some changes, probably it would be valuable to have this change as well.

JamesNK commented on src/System.Text.Json/src/System/Text/Json/Node/JsonObject.cs:
```diff
@@ -526,41 +523,54 @@ public bool TryGetJsonArrayPropertyValue(string propertyName, out JsonArray json
         ///  A collection containing the property values of JSON object.
         /// </summary>
-        public ICollection<JsonNode> PropertyValues => _dictionary.Values;
+        public ICollection<JsonNode> GetPropertyValues() => _dictionary.Values.Select(jsonObjectProperty => jsonObjectProperty.Value).ToList();
```
See previous comment.

JamesNK commented on src/System.Text.Json/src/System/Text/Json/Node/JsonObject.cs:
```diff
@@ -484,6 +437,28 @@ public bool TryGetJsonObjectPropertyValue(string propertyName, out JsonObject js
+        ///  <see langword="true"/> if JSON object property with specified name was found;
+        ///  otherwise, <see langword="false"/>
+        /// </returns>
+        public bool TryGetJsonObjectPropertyValue(string propertyName, out JsonObject jsonObject, StringComparison stringComparison)
```
out parameter should be last

kasiabulat committed:
JsonElement code coverage improved, minor fixes

kasiabulat committed:
review comments included

kasiabulat commented:
I added more tests and addressed new comments. Current coverage:

|Name | Covered | Uncovered | Coverable | Total | Line coverage | Branch coverage|
|---|---|---|---|---|---|---|
| System.Text.Json.JsonArray | 148 | 0 | 148 | 380 | 100% | 100% |   
| System.Text.Json.JsonArrayEnumerator | 10 | 0 | 10 | 50 | 100% | |   
| System.Text.Json.JsonBoolean | 16 | 0 | 16 | 103 | 100% | 100% |   
| System.Text.Json.JsonElement | 629 | 13 | 642 | 2096 | 97.9% |  94.6% |
| System.Text.Json.JsonNode | 263 | 2 | 265 | 598 | 99.2% |  99% |  
| System.Text.Json.JsonNull | 9 | 0 | 9 | 73 | 100% | |
| System.Text.Json.JsonNumber | 75 | 0 | 75 | 622 | 100% |  100% | 
| System.Text.Json.JsonObject | 203 | 0 | 203 | 597 | 100% |  100% |   
| System.Text.Json.JsonObjectEnumerator | 27 | 0 | 27 | 81 | 100% | 100% | 
| System.Text.Json.JsonObjectProperty | 11 | 0 | 11 | 22 | 100% |   |  
| System.Text.Json.JsonString | 27 | 0 | 27 | 212 | 100% | 100% |  

Uncovered lines are either unreachable / unsupported or handling unspecified behavior. 

ahsonkhan commented on src/System.Text.Json/tests/JsonObjectTests.cs:
```diff
@@ -712,11 +654,147 @@ public static void TestArgumentNullValidation()
+
+            Assert.False(jsonObject.ContainsProperty("ENCYCLOPAEDIA"));
+            Assert.False(jsonObject.TryGetPropertyValue("ENCYCLOPAEDIA", out JsonNode jsonNode));
+            Assert.Null(jsonNode);
```
**Question:** Do we have a test where the property we are searching for has the value "null" so TryGet returns true and JsonNode is not null, but rather is a JsonNull? If so, great. Otherwise, let's add it.

kasiabulat replied:
We have such test for indexer, I can add it for Get/(Try)Get as well.

ahsonkhan commented on src/System.Text.Json/tests/JsonObjectTests.cs:
```diff
@@ -712,11 +654,147 @@ public static void TestArgumentNullValidation()
+            Assert.False(jsonObject.TryGetPropertyValue("ENCYCLOPAEDIA", out JsonNode jsonNode));
+            Assert.Null(jsonNode);
+            Assert.Throws<KeyNotFoundException>(() => jsonObject.GetPropertyValue("ENCYCLOPAEDIA"));
+            jsonObject.Remove("ENCYCLOPAEDIA");
```
Assert.False?

ahsonkhan commented on src/System.Text.Json/tests/JsonObjectTests.cs:
```diff
@@ -712,11 +654,147 @@ public static void TestArgumentNullValidation()
+            jsonObject.Remove("ENCYCLOPAEDIA", StringComparison.CurrentCulture);
+            Assert.Equal(4, jsonObject.Count());
+
+            Assert.True(jsonObject.ContainsProperty("ENCYCLOPAEDIA", StringComparison.InvariantCultureIgnoreCase));
```
How about Ordinal and OrdinalIgnoreCase?

ahsonkhan commented on src/System.Text.Json/tests/JsonObjectTests.cs:
```diff
@@ -712,11 +654,147 @@ public static void TestArgumentNullValidation()
+            jsonObject.Remove("ENCYCLOPAEDIA");
+            Assert.Equal(4, jsonObject.Count());
+
+            Assert.False(jsonObject.ContainsProperty("ENCYCLOPAEDIA", StringComparison.CurrentCulture));
```
We should add tests where we set the current culture to something else and then check both cases (using `StringComparison.CurrentCulture` finds what we are looking for because now it matches, and one where because we use currentculture, it doesn't).

ahsonkhan reviewed (commented):
Some feedback, mostly nits and questions.

kasiabulat commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNull.cs:
```diff
@@ -0,0 +1,73 @@
+        ///   Calculates a hash code of this instance.
+        /// </summary>
+        /// <returns>A hash code for this instance.</returns>
+        public override int GetHashCode() => -1;
```
I chose -1 so that it has different hash code than JsonBoolean(false).

kasiabulat commented on src/System.Text.Json/src/System/Text/Json/Node/JsonObject.cs:
```diff
@@ -99,315 +94,195 @@ public void Add(string propertyName, JsonNode propertyValue)
         ///   Provided property name is null.
         /// </exception>
-        public void Add(string propertyName, short propertyValue) => Add(propertyName, new JsonNumber(propertyValue));
+        public bool Remove(string propertyName)
```
I believe `Remove(string propertyName)` is faster. The overload taking `StringComparison` always iterates through the list until it finds the match, while regular remove should be approaching O(1).  So `Remove(string propertyName, StringComparison.Ordinal)` will do the same thing, but slower.

ahsonkhan replied:
Right, exactly.

So, if someone calls `Remove(string propertyName, StringComparison.Ordinal)`, we should internally special case that enum value and call `Remove(string propertyName)`, since it is equivalent and faster.

kasiabulat replied:
Right, sure, I'll change it.

kasiabulat commented on src/System.Text.Json/src/System/Text/Json/Node/JsonObject.cs:
```diff
@@ -99,315 +94,195 @@ public void Add(string propertyName, JsonNode propertyValue)
-        ///   Provided value or property name is null.
-        /// </exception>
-        public void Add(string propertyName, string propertyValue) => Add(propertyName, new JsonString(propertyValue));
+            // Add property to linked list:
```
This is to preserve dictionary order.

ahsonkhan replied:
And ordering can't be preserved with a regular `List<T>`? Or is the point here, that re-adjusting a linked list (to insert somewhere in the middle) is faster than having to slide the entire list over?

kasiabulat replied:
This approach was a suggestion of @bartonjs, so probably he can tell more about the benefits he wanted to achieve. I understood that we want to combine the best traits of dictionary and list - have a fast access to elements and store elements in the same order they were added.

kasiabulat committed:
JsonReaderOptions changed to JsonDocumentOptions, Add(JsonArray) overload removed

kasiabulat commented:
I changed `JsonReaderOptions` into `JsonDocumentOptions` and removed unnecessary Add(string, JsonArray) overload from `JsonObject`, I will address other feedback in subsequent PR.

kasiabulat commented on src/System.Text.Json/ref/System.Text.Json.cs:
```diff
@@ -270,8 +235,42 @@ public abstract partial class JsonNode
+        public static implicit operator System.Text.Json.JsonNode (uint value) { throw null; }
+        [System.CLSCompliantAttribute(false)]
+        public static implicit operator System.Text.Json.JsonNode (ulong value) { throw null; }
+        public static System.Text.Json.JsonNode Parse(string json, System.Text.Json.JsonReaderOptions options = default(System.Text.Json.JsonReaderOptions), System.Text.Json.DuplicatePropertyNameHandling duplicatePropertyNameHandling = System.Text.Json.DuplicatePropertyNameHandling.Replace) { throw null; }
```
Ok, for now I changed it `JsonReaderOptions` into `JsonDocumentOptions`. I am not sure if we should add `DuplicatePropertyNameHandling` to `JsonDocumentOptions`, because `JsonDocument` does not support it, so users will be confused by seeing that setting this option does not change anything it this case.

ahsonkhan replied:
I think it's fine for `JsonDocumentOptions` to throw something like `ArgumentException` while we don't have support for `DuplicatePropertyNameHandling`. Having a separate option being passed-in doesn't scale well. If we want to add a new option that is specific to JsonNode/JsonDocument, would we then add a new Parse overload?

Also, we can re-visit this if JsonDocument cannot support the various `DuplicatePropertyNameHandling` modes, and maybe create a JsonNodeOptions struct that is similar to JsonReaderOptions/JsonDocumentOptions, but with the extra option for duplicates.

kasiabulat replied:
Yes, probably `JsonNodeOptions` would be more appropriate.

joperezr commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNode.cs:
```diff
@@ -31,7 +29,10 @@ public abstract class JsonNode
+        /// <exception cref="ArgumentException">
+        ///   Provided <see cref="JsonElement"/> was not build from <see cref="JsonNode"/>.
+        /// </exception>
+        public static JsonNode GetNode(JsonElement jsonElement) => !jsonElement.IsImmutable ? (JsonNode)jsonElement._parent : throw new ArgumentException(SR.NotNodeJsonElementParent);
```
Did we ever discussed why is this one called GetNode and not GetJsonNode? Giving the full name of the return type would be better in my opinion.

kasiabulat replied:
We didn't discuss it; yes, probably `GetJsonNode` would be better.

joperezr commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNode.Traversal.cs:
```diff
@@ -0,0 +1,327 @@
+                }
+            }
+
+            Debug.Assert(toReturn != null);
```
Is this actually a failing scenario? What happens today if I do JsonNode.DeepCopy(JsonDocument.Parse(""))

kasiabulat replied:
JsonDocument will throw JsonReaderException then.

joperezr commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNode.Traversal.cs:
```diff
@@ -0,0 +1,327 @@
+        /// <returns>JSON representation of current instance.</returns>
+        public string ToJsonString()
+        {
+            var stream = new MemoryStream();
```
I believe we should also wrap this MemoryStream inside a using statement since it is also disposable.

kasiabulat replied:
I changed it to ArrayBufferWriter anyway.

kasiabulat commented on src/System.Text.Json/src/System/Text/Json/Node/JsonArray.cs:
```diff
@@ -215,12 +207,13 @@ public JsonArray(IEnumerable<decimal> values) : this()
             set
             {
-                _list[idx] = value;
+                _list[idx] = value ?? new JsonNull();
```
I can create internal static field for this.

kasiabulat commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNode.Traversal.cs:
```diff
@@ -0,0 +1,327 @@
+                        break;
+                }
+
+                writer.Flush();
```
Ok, sure.

kasiabulat commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNode.Traversal.cs:
```diff
@@ -0,0 +1,327 @@
+            using (var writer = new Utf8JsonWriter(stream))
+            {
+                WriteTo(writer);
+                return JsonHelpers.Utf8GetString(stream.ToArray());
```
Ok, thanks!

kasiabulat commented on src/System.Text.Json/ref/System.Text.Json.cs:
```diff
@@ -270,8 +235,42 @@ public abstract partial class JsonNode
+        public static implicit operator System.Text.Json.JsonNode (System.Guid value) { throw null; }
+        public static implicit operator System.Text.Json.JsonNode (short value) { throw null; }
+        public static implicit operator System.Text.Json.JsonNode (int value) { throw null; }
+        public static implicit operator System.Text.Json.JsonNode (long value) { throw null; }
```
What kind of API do you mean exactly?

ahsonkhan replied:
https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/operator-overloads

> ✓ CONSIDER providing methods with friendly names that correspond to each overloaded operator.

Generally, when you have implicit operators, the recommendation is to have a way to accomplish the same thing but by calling APIs with friendly names (as an alternative). Maybe this could be ctor overloads or static methods on JsonNode.

kasiabulat replied:
I think we do have it - all of the primary types (JsonNumber, JsonString, JsonBoolean) already have such constructors and JsonNumber additionally has getter and setter overloads. Do you want to add them to JsonNode as well? Didn't we want to keep JsonNode's constructor private protected? And JsonNode is an abstract class.

kasiabulat commented on src/System.Text.Json/src/System/Text/Json/Node/JsonObject.cs:
```diff
@@ -526,41 +523,54 @@ public bool TryGetJsonArrayPropertyValue(string propertyName, out JsonArray json
         ///  A collection containing the property values of JSON object.
         /// </summary>
-        public ICollection<JsonNode> PropertyValues => _dictionary.Values;
+        public IReadOnlyCollection<JsonNode> GetPropertyValues() => _dictionary.Values.Select(jsonObjectProperty => jsonObjectProperty.Value).ToList();
```
Because dictionary values are now represented as `JsonObjectProperties` not `JsonNodes`.

kasiabulat commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNode.TraversalHelpers.cs:
```diff
@@ -0,0 +1,81 @@
+            }
+        }
+
+        private struct RecursionStackFrame
```
Ok.

joperezr commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNode.TraversalHelpers.cs:
```diff
@@ -0,0 +1,81 @@
+            ref JsonNode toReturn,
+            DuplicatePropertyNameHandling duplicatePropertyNameHandling = DuplicatePropertyNameHandling.Replace)
+        {
+            if (currentNodes.Any())
```
Can we just add some comment in here saying that currentNodes will really mean that we are inside either an object or an array?

kasiabulat replied:
Ok, I can add it.

kasiabulat commented on src/System.Text.Json/src/System/Text/Json/Node/JsonObject.cs:
```diff
@@ -431,16 +306,72 @@ public JsonNode GetPropertyValue(string propertyName)
+        {
+            if (!TryGetPropertyValue(propertyName, stringComparison, out JsonNode jsonNode))
+            {
+                throw new KeyNotFoundException(SR.Format(SR.PropertyNotFound, propertyName));
```
Ok, I will change it.

kasiabulat commented on src/System.Text.Json/tests/JsonArrayTests.cs:
```diff
@@ -193,8 +191,53 @@ public static void TestCreatingJsonArrayFromStringArray()
+        }
+
+        [Fact]
+        public static void TestCreatingJsonArrayFromIEnumerableOfStrings()
```
In line 213: 

https://github.com/kasiabulat/corefx/blob/c5bf73c3c7b111c9e2369ee26461560dc48fcb01/src/System.Text.Json/tests/JsonArrayTests.cs#L211-L213

ahsonkhan replied:
Ah yes, missed that.

kasiabulat commented on src/System.Text.Json/tests/JsonArrayTests.cs:
```diff
@@ -283,36 +333,117 @@ public static void TestAddingToJsonArray()
+
+            jsonArray[0] = null;
+            Assert.Equal(1, jsonArray.Count());
+            Assert.IsType<JsonNull>(jsonArray[0]);
```
It will pass.

kasiabulat replied:
And, actually, I have this check in line 361.

ahsonkhan replied:
Sounds good. I was interested in what happens when you pass in a a new instance of JsonNull via the ctor (and if reference equality would change the result), but line 361 is implicitly creating one from `null` anyway.

kasiabulat commented on src/System.Text.Json/tests/JsonArrayTests.cs:
```diff
@@ -283,36 +333,117 @@ public static void TestAddingToJsonArray()
+            var jsonArray = new JsonArray() { 1 };
+            Assert.Equal(1, jsonArray.Count);
+
+            jsonArray.Insert(0, 0);
```
I checked it - it complies and JsonNull node is added in runtime.

kasiabulat commented on src/System.Text.Json/tests/JsonElementWithNodeParentTests.cs:
```diff
@@ -0,0 +1,295 @@
+        [Fact]
+        public static void TestGetRawText()
+        {
+            string jsonToParse = @"{""property name"":""value""}";
```
GetRawText will change escaped characters to their unescaped equivalents in that case. It isn't actually returning "raw text" for JsonNode parent, because we don't have it anymore, but rather ToJsonString result (which is pointed out as a remark).

kasiabulat commented on src/System.Text.Json/tests/JsonElementWithNodeParentTests.cs:
```diff
@@ -0,0 +1,295 @@
+                ["array"] = new JsonArray() { 1, 2 }
+            };
+
+            var stream = new MemoryStream();
```
I can change it.

kasiabulat commented on src/System.Text.Json/tests/JsonNullTests.cs:
```diff
@@ -0,0 +1,99 @@
+
+            Assert.True(jsonNull1.Equals(jsonNull2));
+            Assert.True(jsonNull1 == jsonNull2);
+            Assert.False(jsonNull1 != jsonNull2);
```
I am not sure what you mean. Should the last line be:
```csharp
Assert.True(jsonString1 == jsonString2);
```

Yes, `new JsonString("1")` is equal to `new JsonString("1")`, because of overloaded equals comparing values. For JsonNumbers I am comparing actual value stored as string, so for example `new JsonNumber("1")` is equal to `new JsonNumber("1")` as well.

ahsonkhan replied:
> Should the last line be

Yep, that was a typo.

The behavior makes sense to me. Thanks for clarifying.

kasiabulat commented on src/System.Text.Json/tests/JsonNullTests.cs:
```diff
@@ -0,0 +1,99 @@
+            // Only for null JsonNode / different derived type this will return false:
+
+            JsonNode jsonNodeNull = null;
+            Assert.False(jsonNull1.Equals(jsonNodeNull));
```
Having JsonNull not equal to JsonNode null reference and other derived types null references makes it possible to differentiate between such cases:

```csharp
  JsonNode node = jsonObject["value"] as JsonNull;
  if(node == null)
  {
    // cast failed, jsonObject["value"] is not of type JsonNull
  }
  if(node != null)
  {
    // cast succeeded, jsonObject["value"] is a JsonNull instance
  }
```

I don't know how to fix "Assert.True(jsonNull == null);" case, but probably it's OK?

kasiabulat commented on src/System.Text.Json/tests/JsonStringTests.cs:
```diff
@@ -20,7 +20,7 @@ public static void TestDefaultCtor()
         [InlineData("value with some spaces")]
         [InlineData("     leading spaces")]
-        [InlineData("trailing spaces"     )]
+        [InlineData("trailing spaces")]
```
Thanks, I don't know how it happened.

kasiabulat commented on src/System.Text.Json/ref/System.Text.Json.cs:
```diff
@@ -364,48 +348,35 @@ public sealed partial class JsonNumber : System.Text.Json.JsonNode, System.IEqua
+        public System.Text.Json.JsonNode GetPropertyValue(string propertyName, System.StringComparison stringComparison) { throw null; }
+        public System.Collections.Generic.IReadOnlyCollection<System.Text.Json.JsonNode> GetPropertyValues() { throw null; }
         public bool Remove(string propertyName) { throw null; }
+        public bool Remove(string propertyName, System.StringComparison stringComparison) { throw null; }
```
We do have such tests, for all overloads, but not for all enum values. I can improve this later on.

iSazonov commented:
Is there a nuget with the new APIs?

ahsonkhan commented:
> Is there a nuget with the new APIs?

Yes, you can grab the alpha release of the `System.Text.Json` NuGet package from the daily feed and add it as a Package Reference to your project (as long as the project is netstandard2.0 or higher):
https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet5/nuget/v3/index.json
Version: 5.0.0-alpha.1.19563.6 (or higher)

You could also grab the nightly shared framework, if you are targeting .NET Core 5.0:
https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/docs/project/dogfooding.md#install-prerequisites
